includes/class-wcsatt-product.php

```diff
@@ -0,0 +1,856 @@
+			if ( self::is_subscription( $product ) ) {
+
+				if ( $switched_scheme ) {
+					// No infinite loops, thank you.
+					WCS_ATT_Scheme_Prices::remove_price_html_filters();
+					$price_html = $product->get_price_html();
+					WCS_ATT_Scheme_Prices::add_price_html_filters();
+				}
+
+				$args[ 'price' ] = $price_html;
+
+				$price_html = WC_Subscriptions_Product::get_price_string( $product, $args );
+
+			// Subscription state is undefined? Construct a special price string.
+			} elseif ( is_null( $scheme_key ) ) {
+
+				$schemes         = self::get_subscription_schemes( $product );
+				$base_scheme     = WCS_ATT_Scheme_Prices::get_base_scheme( $product );
+				$base_scheme_key = $base_scheme->get_key();
+
+				// Temporarily apply base scheme on product object.
+				$switched_scheme = self::set_subscription_scheme( $product, $base_scheme_key );
+
+				if ( $product->is_type( 'variable' ) && $product->get_variation_price( 'min' ) !== $product->get_variation_price( 'max' ) ) {
+					$has_variable_price = true;
+				} elseif (  $product->is_type( 'bundle' ) && $product->get_bundle_price( 'min' ) !== $product->get_bundle_price( 'max' ) ) {
+					$has_variable_price = true;
+				} elseif ( $product->is_type( 'composite' ) && $product->get_composite_price( 'min' ) !== $product->get_composite_price( 'max' ) ) {
+					$has_variable_price = true;
+				} else {
+					$has_variable_price = false;
+				}
+
+				$html_from_text = WCS_ATT_Core_Compatibility::get_price_html_from_text( $product );
+
+				if ( self::has_forced_subscription( $product ) ) {
+
+					// No infinite loops, thank you.
+					WCS_ATT_Scheme_Prices::remove_price_html_filters();
+					$price_html = $product->get_price_html();
+					WCS_ATT_Scheme_Prices::add_price_html_filters();
+
+					$args[ 'price' ] = $price_html;
+
+					$price_html = WC_Subscriptions_Product::get_price_string( $product, $args );
+
+					if ( ( $has_variable_price || sizeof( $schemes ) > 1 ) && false === strpos( $price_html, $html_from_text ) ) {
+						$price_html = sprintf( _x( '%1$s%2$s', 'Price range: from', WCS_ATT::TEXT_DOMAIN ), $html_from_text, $price_html );
+					}
+
+				} else {
+
+					$suffix_price_html        = '';
+					$use_discount_html_format = false;
+					$has_variable_discount    = false;
+					$discount                 = '';
+
+					// Show discount format if all schemes are of the 'inherit' pricing mode type.
+					$price_filter_exists = WCS_ATT_Scheme_Prices::price_filter_exists( $schemes );
+
+					if ( $price_filter_exists ) {
+						foreach ( $schemes as $scheme ) {
+							if ( $scheme->has_price_filter() && 'inherit' === $scheme->get_pricing_mode() ) {
+								$use_discount_html_format = true;
+								if ( $discount !== $scheme->get_discount() ) {
+									if ( '' === $discount ) {
+										$discount = $scheme->get_discount();
+									} else {
+										$has_variable_discount = true;
+									}
+								}
+							}
+						}
```

Comment:
This basically loops through all schemes to decide two things:

1. Whether to use a "discount-based" html format.
2. Whether different schemes have varying discounts.

We could create 2 dedicated functions for these, but they wouldn't be useful in any other place.

Large, deeply nested functions are evil indeed. Maybe it's just me and my scripting language background, but string-generating functions like this one are a typical example of something that could become more complex to review/understand if you split it up too much or use objects to do the same thing.